dotnet-engine/Yggdrasil.Engine/FFI.cs

```diff
@@ -0,0 +1,152 @@
+
+    private static readonly NewEngineDelegate _newEngine;
+    private static readonly FreeEngineDelegate _freeEngine;
+    private static readonly GetMetricsDelegate _getMetrics;
+    private static readonly TakeStateDelegate _take_state;
+    private static readonly CheckEnabledDelegate _check_enabled;
+    private static readonly CheckVariantDelegate _check_variant;
+    private static readonly FreeResponseDelegate _free_response;
+    private static readonly CountToggleDelegate _count_toggle;
+    private static readonly CountVariantDelegate _count_variant;
+
+    static FFI()
+    {
+        string dllPath = GetLibraryPath();
+        IntPtr libHandle = NativeLibrary.Load(dllPath);
+
+        _newEngine = Marshal.GetDelegateForFunctionPointer<NewEngineDelegate>(
+            NativeLibrary.GetExport(libHandle, "new_engine")
+        );
+
+        _freeEngine = Marshal.GetDelegateForFunctionPointer<FreeEngineDelegate>(
+            NativeLibrary.GetExport(libHandle, "free_engine")
+        );
+
+        _getMetrics = Marshal.GetDelegateForFunctionPointer<GetMetricsDelegate>(
+            NativeLibrary.GetExport(libHandle, "get_metrics")
+        );
+
+        _take_state = Marshal.GetDelegateForFunctionPointer<TakeStateDelegate>(
+            NativeLibrary.GetExport(libHandle, "take_state")
+        );
+
+        _check_enabled = Marshal.GetDelegateForFunctionPointer<CheckEnabledDelegate>(
+            NativeLibrary.GetExport(libHandle, "check_enabled")
+        );
+
+        _check_variant = Marshal.GetDelegateForFunctionPointer<CheckVariantDelegate>(
+            NativeLibrary.GetExport(libHandle, "check_variant")
+        );
+
+        _free_response = Marshal.GetDelegateForFunctionPointer<FreeResponseDelegate>(
+            NativeLibrary.GetExport(libHandle, "free_response")
+        );
+
+        _count_toggle = Marshal.GetDelegateForFunctionPointer<CountToggleDelegate>(
+            NativeLibrary.GetExport(libHandle, "count_toggle")
+        );
+
+        _count_variant = Marshal.GetDelegateForFunctionPointer<CountVariantDelegate>(
+            NativeLibrary.GetExport(libHandle, "count_variant")
+        );
+    }
+
+    private static string GetLibraryPath()
+    {
+        string libraryName;
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            libraryName = "libyggdrasilffi.dll";
+        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            libraryName = "libyggdrasilffi.so";
+        else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            libraryName = "libyggdrasilffi.dylib";
+        else
+            throw new PlatformNotSupportedException("Unsupported platform");
+
+        string assemblyLocation = Assembly.GetExecutingAssembly().Location;
+        string assemblyDirectory = Path.GetDirectoryName(assemblyLocation);
+
+        return Path.Combine(assemblyDirectory, libraryName);
```

Comment:
To make the tests pick up the execution location so they can find the so/dll/dylib